From daa4ad3de6ec06b43e3aeaa3aa979b159466a869 Mon Sep 17 00:00:00 2001
From: Alex James <alex.james@broadcom.com>
Date: Wed, 29 Apr 2026 22:57:06 +0000
Subject: [PATCH 4/7] x86/coco: Move snp_dmi_setup function to common code

Commit 0f4a1e80989a ("x86/sev: Skip ROM range scans and validation for
SEV-SNP guests") introduced a SEV-SNP-specific dmi_setup function to
only scan for DMI tables when the corresponding EFI configuration tables
are available. The fallback path which manually probes for the DMI
tables may otherwise access ROM regions which have not been
prevalidated.

TDX guests also require encrypted memory to be accepted before it can be
read. Move the existing snp_dmi_setup() function to the common x86 coco
code so it can reused on TDX.

Signed-off-by: Alex James <alex.james@broadcom.com>
---
 arch/x86/coco/core.c          | 13 +++++++++++++
 arch/x86/coco/sev/core.c      | 13 -------------
 arch/x86/include/asm/coco.h   |  2 ++
 arch/x86/mm/mem_encrypt_amd.c |  4 ++--
 4 files changed, 17 insertions(+), 15 deletions(-)

diff --git a/arch/x86/coco/core.c b/arch/x86/coco/core.c
index 0f81f70aca82..d5606a8051bb 100644
--- a/arch/x86/coco/core.c
+++ b/arch/x86/coco/core.c
@@ -10,6 +10,8 @@
 
 #include <linux/export.h>
 #include <linux/cc_platform.h>
+#include <linux/dmi.h>
+#include <linux/efi.h>
 #include <linux/string.h>
 #include <linux/random.h>
 
@@ -204,6 +206,17 @@ void cc_platform_set(enum cc_attr attr)
 	}
 }
 
+/*
+ * SEV-SNP and TDX guests should only execute dmi_setup() if EFI_CONFIG_TABLES
+ * are enabled, as the alternative (fallback) logic for DMI probing in the
+ * legacy ROM region can cause a crash since this region is not pre-validated.
+ */
+__init void cc_dmi_setup(void)
+{
+	if (efi_enabled(EFI_CONFIG_TABLES))
+		dmi_setup();
+}
+
 __init void cc_random_init(void)
 {
 	/*
diff --git a/arch/x86/coco/sev/core.c b/arch/x86/coco/sev/core.c
index 2be730765f83..2c66103a704a 100644
--- a/arch/x86/coco/sev/core.c
+++ b/arch/x86/coco/sev/core.c
@@ -19,11 +19,9 @@
 #include <linux/kernel.h>
 #include <linux/mm.h>
 #include <linux/cpumask.h>
-#include <linux/efi.h>
 #include <linux/platform_device.h>
 #include <linux/io.h>
 #include <linux/psp-sev.h>
-#include <linux/dmi.h>
 #include <uapi/linux/sev-guest.h>
 
 #include <asm/init.h>
@@ -2325,17 +2323,6 @@ void __head __noreturn snp_abort(void)
 	sev_es_terminate(SEV_TERM_SET_GEN, GHCB_SNP_UNSUPPORTED);
 }
 
-/*
- * SEV-SNP guests should only execute dmi_setup() if EFI_CONFIG_TABLES are
- * enabled, as the alternative (fallback) logic for DMI probing in the legacy
- * ROM region can cause a crash since this region is not pre-validated.
- */
-void __init snp_dmi_setup(void)
-{
-	if (efi_enabled(EFI_CONFIG_TABLES))
-		dmi_setup();
-}
-
 static void dump_cpuid_table(void)
 {
 	const struct snp_cpuid_table *cpuid_table = snp_cpuid_get_table();
diff --git a/arch/x86/include/asm/coco.h b/arch/x86/include/asm/coco.h
index aa6c8f8ca958..4d7035dbeeae 100644
--- a/arch/x86/include/asm/coco.h
+++ b/arch/x86/include/asm/coco.h
@@ -22,6 +22,7 @@ static inline void cc_set_mask(u64 mask)
 
 u64 cc_mkenc(u64 val);
 u64 cc_mkdec(u64 val);
+void cc_dmi_setup(void);
 void cc_random_init(void);
 #else
 #define cc_vendor (CC_VENDOR_NONE)
@@ -36,6 +37,7 @@ static inline u64 cc_mkdec(u64 val)
 {
 	return val;
 }
+static inline void cc_dmi_setup(void) { }
 static inline void cc_random_init(void) { }
 #endif
 
diff --git a/arch/x86/mm/mem_encrypt_amd.c b/arch/x86/mm/mem_encrypt_amd.c
index 86a476a426c2..798d2366b7b9 100644
--- a/arch/x86/mm/mem_encrypt_amd.c
+++ b/arch/x86/mm/mem_encrypt_amd.c
@@ -505,10 +505,10 @@ void __init sme_early_init(void)
 		/*
 		 * DMI setup behavior for SEV-SNP guests depends on
 		 * efi_enabled(EFI_CONFIG_TABLES), which hasn't been
-		 * parsed yet. snp_dmi_setup() will run after that
+		 * parsed yet. cc_dmi_setup() will run after that
 		 * parsing has happened.
 		 */
-		x86_init.resources.dmi_setup = snp_dmi_setup;
+		x86_init.resources.dmi_setup = cc_dmi_setup;
 	}
 
 	/*
-- 
2.53.0

